transport/conn.go

```diff
@@ -44,13 +44,58 @@ type ConnMetrics struct {
 	InQueue  atomic.Uint32
 }
 
+type coalescingStrategy struct {
+	gaps  [16]time.Duration
+	pos   int
+	total time.Duration
+
+	last  time.Time
+	ready bool
+}
+
+const (
+	maxCoalesceWaitTime  = time.Millisecond
+	maxCoalescedRequests = 100
+)
+
+func (cs *coalescingStrategy) apply() {
+	if cs.ready {
+		waitTime := 2 * cs.total / 16
+		if waitTime > maxCoalesceWaitTime {
+			waitTime = maxCoalesceWaitTime
+		}
+
+		time.Sleep(waitTime)
+	}
+}
+
+func (cs *coalescingStrategy) update() {
+	if cs.last.IsZero() {
+		cs.last = time.Now()
+		return
+	}
+
+	gap := time.Since(cs.last)
+	cs.last = time.Now()
+	cs.total += gap - cs.gaps[cs.pos]
+	cs.gaps[cs.pos] = gap
```

Comment:
I'm not sure about this code
- it looks like the array is nit needed and you could do with the last value only
- it looks like in L80 we calculate difference between current and last batch